### ISSUE TITLE
Narrowing of enum types can cause false positives after mutating method calls

### ISSUE THREAD

**Issue by dima179-fuse** [python/mypy#21604]
**Bug Report**

Mypy preserves a narrowing of a mutable attribute across a method call that can mutate that attribute.

In the example below, `d.e` is narrowed to `Literal[E.B]` after checking `d.e != E.A`. However, after calling `d.set_e(e)`, where `e: E` may be either `E.A` or `E.B`, mypy continues to treat `d.e` as `Literal[E.B]`.

This appears unsound because `set_e` can assign `E.A` to `d.e`, making the subsequent `if d.e == E.A` branch reachable at runtime.

**To Reproduce**

```python
from enum import StrEnum, auto
from dataclasses import dataclass


class E(StrEnum):
    A = auto()
    B = auto()


@dataclass
class D:
    e: E

    def set_e(self, e: E) -> None:
        self.e = e


def func(d: D, e: E) -> None:
    reveal_type(d.e)

    if d.e == E.A:
        return

    d.set_e(e)

    reveal_type(d.e)

    if d.e == E.A:
        return

    reveal_type(d.e)
```

**Expected Behavior**

After the call to `d.set_e(e)`, mypy should invalidate the previous narrowing of `d.e` and treat it as `E`, since the method can mutate the attribute.

Expected reveals:

```text
reveal_type(d.e)  # E
...
reveal_type(d.e)  # E
...
reveal_type(d.e)  # Literal[E.B]
```

The second `if d.e == E.A` should not be reported as a non-overlapping comparison.

For example, the following call can make `d.e` equal to `E.A` after the method call:

```python
d = D(E.B)
func(d, E.A)
```

**Actual Behavior**

```text
main.py:16: note: Revealed type is "__main__.E"
main.py:20: note: Revealed type is "Literal[__main__.E.B]"
main.py:21: error: Non-overlapping equality check (left operand type: "Literal[E.B]", right operand type: "Literal[E.A]")  [comparison-overlap]
main.py:22: error: Statement is unreachable  [unreachable]
main.py:23: note: Revealed type is "Literal[__main__.E.B]"
Found 2 errors in 1 file (checked 1 source file)
```

The reported unreachable branch is reachable at runtime.

**Your Environment**

- Mypy version used: 2.1.0
- Mypy command-line flags: --strict-equality --strict-equality-for-none --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.13
Gist URL: https://gist.github.com/mypy-play/b5db4d6e79381480dac2e843d472a508
Playground URL: https://mypy-play.net/?mypy=2.1.0&python=3.13&flags=strict-equality%2Cstrict-equality-for-none%2Cwarn-unreachable&gist=b5db4d6e79381480dac2e843d472a508

**Comment by ilevkivskyi** [comment]
This is a intentional trade-off, and there is not much we can do about it. (Also I am quite sure we already have few duplicates of this, but I am too lazy to search.)

**Comment by dima179-fuse** [comment]
@ilevkivskyi what do you mean an intentional trade-off???
This seems to make that narrowing of enums to `Literal[something]` rather useless

**Comment by JukkaL** [comment]
I think narrowing enums to literals in particular causes some friction in tests. This may be something we could possibly tweak, though it would be a complex tradeoff to change the behavior now. We can reopen this issue, but focused on enum narrowing.

**Comment by JukkaL** [comment]
Reopened with a more specific issue title.

**Comment by ilevkivskyi** [comment]
@JukkaL If you want to work on this, please find the original issue.

**Comment by lheckemann** [comment]
https://github.com/python/mypy/issues/9457 is the original issue, I guess?

**Comment by ilevkivskyi** [comment]
Yes, to be more precise that one was selected as a canonical representative of this class of duplicates.
